subsys/fs/ext2/ext2_impl.c

```diff
@@ -0,0 +1,1558 @@
+/*
+ * Copyright (c) 2023 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/init.h>
+#include <zephyr/fs/fs.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/util.h>
+
+#include "ext2.h"
+#include "ext2_impl.h"
+#include "ext2_struct.h"
+#include "ext2_diskops.h"
+#include "ext2_bitmap.h"
+
+LOG_MODULE_REGISTER(ext2, CONFIG_EXT2_LOG_LEVEL);
+
+static struct ext2_data __fs;
+static bool initialized;
+
+/* Structures for blocks slab alocator */
+struct k_mem_slab ext2_block_memory_slab, ext2_block_struct_slab;
+char __aligned(8) __ext2_block_memory_buffer[CONFIG_EXT2_BLOCKS_BUF_SIZE];
+char __aligned(8) __ext2_block_struct_buffer[CONFIG_EXT2_BLOCKS_BUF_SIZE / 64];
+
+/* Initialize heap memory allocator */
+K_HEAP_DEFINE(ext2_heap, CONFIG_EXT2_HEAP_SIZE);
+
+/* Helper functions --------------------------------------------------------- */
+
+void error_beharior(struct ext2_data *fs, const char *msg)
+{
+	LOG_ERR("File system corrupted: %s", msg);
+
+	/* If file system is not initialized panic */
+	if (!fs->sblock) {
+		LOG_ERR("File system data not found. Panic...");
+		k_panic();
+	}
+
+	switch (EXT2_DATA_SBLOCK(fs)->s_errors) {
+	case EXT2_ERRORS_CONTINUE:
+		/* Do nothing */
+		break;
+	case EXT2_ERRORS_RO:
+		LOG_WRN("Marking file system as read only");
+		fs->flags |= EXT2_DATA_FLAGS_RO;
+		break;
+	case EXT2_ERRORS_PANIC:
+		LOG_ERR("Panic...");
+		k_panic();
+		break;
+	default:
+		LOG_ERR("Unrecognized errors behavior in superblock s_errors field. Panic...");
+		k_panic();
+	}
+}
+
+void *ext2_heap_alloc(size_t size)
+{
+	void *ptr = k_heap_alloc(&ext2_heap, size, K_NO_WAIT);
+
+	if (!ptr) {
+		return NULL;
+	}
+	return ptr;
+}
+
+void  ext2_heap_free(void *ptr)
+{
+	k_heap_free(&ext2_heap, ptr);
+}
+
+/* Block operations --------------------------------------------------------- */
+
+static struct ext2_block *get_block_struct(void)
+{
+	int ret;
+	struct ext2_block *b;
+
+	ret = k_mem_slab_alloc(&ext2_block_struct_slab, (void **)&b, K_NO_WAIT);
+	if (ret < 0) {
+		LOG_ERR("get block: alloc block struct error %d", ret);
+		return NULL;
+	}
+
+	ret = k_mem_slab_alloc(&ext2_block_memory_slab, (void **)&b->data, K_NO_WAIT);
+	if (ret < 0) {
+		LOG_ERR("get block: alloc block memory error %d", ret);
+		k_mem_slab_free(&ext2_block_struct_slab, (void **)&b);
+		return NULL;
+	}
+	return b;
+}
+
+struct ext2_block *ext2_get_block(struct ext2_data *fs, uint32_t block)
+{
+	int ret;
+	struct ext2_block *b = get_block_struct();
+
+	if (!b) {
+		return NULL;
+	}
```

Comment:
Actually I think, there won't be any significant reduce of checks that should be performed. After proposed change, there still might be a read error, so error check is needed. Maybe that is better because then we will have more control over what error we met. I will consider this change.